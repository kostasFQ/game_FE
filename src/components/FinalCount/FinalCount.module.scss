.finalCount {
  &__container {
    height: 80vh;
    display: flex;
    flex-direction: column;
    justify-content: flex-end;
    align-items: center;

    &__result {
      width: 100%;
      display: flex;
      flex-direction: column;
      align-items: center;
      height: 60%;
      margin-bottom: 30px;

      &_text {
        font-family: 'Otribon', sans-serif;
      }
      
      &_digits {
        display: flex;
        align-items: center;
        font-family: 'Saira', sans-serif;
        height: 90%;
        font-size: 35vh;
        font-weight: bolder;
        color: rgba(253, 86, 86, 0.527);
      }
    }

    &__form {
      width: 90%;
      display: flex;
      flex-direction: column;
      align-items: center;

      &_input {
        text-align: center;
        font-size: 1.5em;
        border-top: none; 
        border-left: none; 
        border-right: none;
        font-family: 'Otribon', sans-serif;
        outline: none;


        &_error {
          border-bottom: solid red 1px;
          &::placeholder {
            color: rgba(255, 0, 0, 0.548);
          }
        }
      }
    }
  }
}