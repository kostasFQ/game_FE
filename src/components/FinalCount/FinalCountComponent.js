import React, { PureComponent } from 'react';
import { withRouter } from 'react-router-dom';
import cn from 'classnames';
import PropTypes from 'prop-types';
import Button from 'components/Buttons/Button';
import buttonStyles from 'components/Buttons/Buttons.module.scss';
import styles from './FinalCount.module.scss';

class FinalCount extends PureComponent {
  state = { value: '', error: false, errorText: undefined };

  restart = () => {
    const { saveCount, toggleGameOver, toggleGameStarted } = this.props;
    saveCount();
    toggleGameOver(false);
    toggleGameStarted(false);
  }

  submitForm = async e => {
    try {
      e.preventDefault();
      const { game: { totalCount, initialTime }, saveResult, history, setUserPlace, setUserName } = this.props;
      const { value } = this.state;
      const body = {
        name: value,
        score: Number(totalCount),
        average: Number((totalCount / initialTime).toFixed(1)),
        seconds: Number(initialTime)
      }
      
      const { data: { error, response } } = await saveResult(body);

      if (error) {
        this.setState(() => ({ error: true, errorText: error }));
        return;
      }
      
      setUserPlace(response);
      setUserName(value);
      this.setState(() => ({ error: false, errorText: undefined }));
      history.push('/');
    } catch (err) {
      console.log(err.message);
    }
  }

  fillField = e => {
    const { value } = e.target;
    this.setState(() => ({ value }));
  }

  render() {
    const { game: { totalCount, initialTime }, history } = this.props;
    const { error, errorText, value } = this.state;

    return (
      <div className={styles.finalCount__container}>
        <div className={styles.finalCount__container__result}>
          <span className={styles.finalCount__container__result_text}>You have</span>
          <div className={styles.finalCount__container__result_digits}>{!! totalCount && totalCount}</div>
          <span className={styles.finalCount__container__result_text}>
            {!!totalCount && totalCount.toString().match(/^1$/) ? 'point ' : 'points '}
            for&nbsp;
            {!!initialTime && initialTime} {!!initialTime && initialTime.toString().match(/^1$/) ? ' second' : ' seconds'}
          </span>
          <div className={styles.finalCount__container__result_text}>it's {(totalCount / initialTime).toFixed(1)} clicks per second</div>
        </div>
        <form method='post' onSubmit={this.submitForm} className={styles.finalCount__container__form}>
          so, you can enter your name
          <input placeholder="" onChange={this.fillField} value={value}
            className={cn(styles.finalCount__container__form_input,
              { [styles.finalCount__container__form_input_error]: error }
            )}
          />
          {!!errorText && <span className={styles.finalCount__container__form_errorMessage}>{errorText}</span>}
          - and -
          <Button type='submit' title='SUBMIT' className={buttonStyles.submitButton} />
        </form>
        - or -
        <Button onClick={this.restart} title='RESTART' className={buttonStyles.restartButton} />
        <Button onClick={() => { history.push('/') }} title='BACK TO MAIN' className={buttonStyles.restartButton} />
      </div>
    )
  }
};

FinalCount.defaultProps = {
  totalCount: 0,
  initialTime: 15
};

FinalCount.propTypes = {
  totalCount: PropTypes.number,
  saveCount: PropTypes.func.isRequired,
  toggleGameOver: PropTypes.func.isRequired,
  toggleGameStarted: PropTypes.func.isRequired,
  history: PropTypes.objectOf(PropTypes.any).isRequired
}

export default withRouter(FinalCount);
