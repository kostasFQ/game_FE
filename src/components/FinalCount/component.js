import React, { PureComponent } from 'react';
import { withRouter } from 'react-router-dom';
import cn from 'classnames';
import PropTypes from 'prop-types';
import Button from 'components/Buttons/Button/component';
import buttonStyles from 'components/Buttons/Buttons.module.scss';
import styles from './FinalCount.module.scss';

class FinalCount extends PureComponent {
  state = { value: '', error: false, errorText: undefined };

  restart = () => {
    const { saveCount, toggleGameOver, toggleGameStarted } = this.props;
    saveCount();
    toggleGameOver(false);
    toggleGameStarted(false);
  }

  submitForm = async e => {
    try {
      e.preventDefault();
      const { game: { totalCount = 0, initialTime }, makeCall, history, setUserPlace, setUserName, saveResultUrl } = this.props;
      const { value } = this.state;
      const body = {
        name: value,
        score: Number(totalCount),
        average: Number((totalCount / initialTime).toFixed(1)),
        seconds: Number(initialTime)
      }
      const { data: { error, response } } = await makeCall(saveResultUrl(), body);

      if (error) {
        this.setState(() => ({ error: true, errorText: error }));
        return;
      }
      setUserPlace(response);
      setUserName(value);
      this.setState(() => ({ error: false, errorText: undefined }));
      history.push('/');
    } catch (err) {
      console.log(err.message);
    }
  }

  fillField = e => {
    const { value } = e.target;
    this.setState(() => ({ value }));
  }

  render() {
    const { game: { totalCount = 0, initialTime }, history } = this.props;
    const { error, errorText } = this.state;

    return (
      <div className={styles.finalCount__container}>
        <div className={styles.finalCount__container__result}>
          <span className={styles.finalCount__container__result_text}>You have</span>
          <div className={styles.finalCount__container__result_digits}>{totalCount}</div>
          <span className={styles.finalCount__container__result_text}>
            {totalCount.toString().match(/^1$/) ? 'point ' : 'points '}
            for&nbsp;
            {initialTime} {initialTime.toString().match(/^1$/) ? ' second' : ' seconds'}
          </span>
          <div className={styles.finalCount__container__result_text}>it's {(totalCount / initialTime).toFixed(1)} clicks per second</div>
        </div>
        <form onSubmit={this.submitForm} className={styles.finalCount__container__form}>
          so, you can enter your name
          <input placeholder="" onChange={this.fillField}
            className={cn(styles.finalCount__container__form_input,
              { [styles.finalCount__container__form_input_error]: error }
            )}
          />
          {!!errorText && <span className={styles.finalCount__container__form_errorMessage}>{errorText}</span>}
          - and -
          <Button type='submit' title='SUBMIT' className={buttonStyles.submitButton} />
        </form>
        - or -
        <Button onClick={this.restart} title='RESTART' className={buttonStyles.restartButton} />
        <Button onClick={() => { history.push('/') }} title='BACK TO MAIN' className={buttonStyles.restartButton} />
      </div>
    )
  }
};

FinalCount.defaultProps = {
  totalCount: 0,
};

FinalCount.propTypes = {
  totalCount: PropTypes.number,
  saveCount: PropTypes.func.isRequired,
  saveResultUrl: PropTypes.func.isRequired,
  toggleGameOver: PropTypes.func.isRequired,
  toggleGameStarted: PropTypes.func.isRequired,
  history: PropTypes.objectOf(PropTypes.any).isRequired
}

export default withRouter(FinalCount);
