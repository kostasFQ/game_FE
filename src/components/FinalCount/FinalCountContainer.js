import { connect } from 'react-redux';
import FinalCount from './FinalCountComponent';
import { saveCount, toggleGameStarted, toggleGameOver, setUserPlace, setUserName } from 'store/game/actions';
import { saveResultUrl } from 'api/urls';
import makeCall from 'api/call';

const mapDispatchToProps = dispatch => ({
  saveCount: count => dispatch(saveCount(count)),
  toggleGameStarted: bool => dispatch(toggleGameStarted(bool)),
  toggleGameOver: bool => dispatch(toggleGameOver(bool)),
  setUserPlace: place => dispatch(setUserPlace(place)),
  setUserName: name => dispatch(setUserName(name)),
  saveResult: value => makeCall(saveResultUrl(saveResultUrl), value)
})

const mapStateToProps = state => ({
  game: state.game,
});

export default connect(mapStateToProps, mapDispatchToProps)(FinalCount);