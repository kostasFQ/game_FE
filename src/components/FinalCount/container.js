import { connect } from 'react-redux';
import FinalCount from './component';
import { saveCount, toggleGameStarted, toggleGameOver } from 'store/game/actions';
import makeCall from 'api/call';

const mapDispatchToProps = dispatch => ({
  saveCount: count => dispatch(saveCount(count)),
  toggleGameStarted: bool => dispatch(toggleGameStarted(bool)),
  toggleGameOver: bool => dispatch(toggleGameOver(bool)),
  makeCall: (url, value) => makeCall(url, value)
})

const mapStateToProps = state => ({
  game: state.game,
});

export default connect(mapStateToProps, mapDispatchToProps)(FinalCount);