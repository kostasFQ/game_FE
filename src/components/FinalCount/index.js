import FinalCount from './FinalCountContainer';

export default FinalCount;