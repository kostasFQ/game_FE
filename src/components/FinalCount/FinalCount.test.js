import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { StaticRouter } from 'react-router-dom';
import _store from '../../../mocks/store';
import FinalCount from './container';
const mockStore = configureStore([thunk]);
const initialState = { game: _store };
const store = mockStore(initialState);

const instance = renderer.create(
  <Provider store={store}>
    <StaticRouter location="/game">
      <FinalCount />
    </StaticRouter>
  </Provider>,
);

test('FinalCount should be render', () => {
  let component = instance.toJSON();
  expect(component).toMatchSnapshot();
})

test('Button "BACK TO MAIN" click', () => {
  const testInstance = instance.root;
  const buttons = testInstance.findAllByType('button');
  const backButton = buttons.find(i => i.props.children === 'BACK TO MAIN');
  backButton.props.onClick();
})

test('Button "RESTART" click', () => {
  const testInstance = instance.root;
  const buttons = testInstance.findAllByType('button');
  const restartButton = buttons.find(i => i.props.children === 'RESTART');
  restartButton.props.onClick();
})

test('Button "SUBMIT" click', () => {
  const testInstance = instance.root;
  const buttons = testInstance.findAllByType('button');
  const submitButton = buttons.find(i => i.props.children === 'SUBMIT');
  submitButton.props.onClick();
})

test('Inputs value was change', () => {
  const testInstance = instance.root;
  const input = testInstance.findByType('input');
  const e = { target: { value: 'test' } };
  input.props.onChange(e);
})

test('Submitting form', () => {
  const testInstance = instance.root;
  const form = testInstance.findByType('form');
  const e = { preventDefault: () => {} };
  form.props.onSubmit(e);
})

test('Catch error', () => {
  const testInstance = instance.root;
  const mockFoo = jest.fn( x => new Error())
  const form = testInstance.findByType('form');
  form.props.onSubmit(mockFoo);
})