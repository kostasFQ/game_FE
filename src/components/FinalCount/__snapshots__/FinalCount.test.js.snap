// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FinalCount should be render 1`] = `
<div
  className="finalCount__container"
>
  <div
    className="finalCount__container__result"
  >
    <span
      className="finalCount__container__result_text"
    >
      You have
    </span>
    <div
      className="finalCount__container__result_digits"
    />
    <span
      className="finalCount__container__result_text"
    >
      points 
      forÂ 
      15
       
       seconds
    </span>
    <div
      className="finalCount__container__result_text"
    >
      it's 
      0.0
       clicks per second
    </div>
  </div>
  <form
    className="finalCount__container__form"
    onSubmit={[Function]}
  >
    so, you can enter your name
    <input
      className="finalCount__container__form_input"
      onChange={[Function]}
      placeholder=""
    />
    - and -
    <button
      className="submitButton"
      onClick={[Function]}
      type="submit"
    >
      SUBMIT
    </button>
  </form>
  - or -
  <button
    className="restartButton"
    onClick={[Function]}
  >
    RESTART
  </button>
  <button
    className="restartButton"
    onClick={[Function]}
  >
    BACK TO MAIN
  </button>
</div>
`;
